### DISCUSSION TITLE
set precision/accuracy on linear coords

### DISCUSSION THREAD

**Discussion by atravert** [spectrochempy/spectrochempy#398] [Developer discussions]
linear coords gernerally returns many decimals, generally beyond reasonable because there is no rounding when `data` are calculated. It would be reasonable  to round the data values:
- the default should be based on the number of decimals of `start`/`stop` values. To be conservative (in terms of accuracy) we could set it to the max number of decimal of these values (e.g. if `start` has 2 decimals and `stop` has 5 decimals, we use 5 decimals) 
- one should also be able to pass a `decimals` parameter to LineraCoord() in order to change the default. 

Typically in omnic files, the `start` value has ~ 4-5 decimals , the `stop` value has ~7-9 decimals, but the display/reports show 3 decimals (which is very reasonable in my view). So in read_omnic we woulmd set `LinearCoord(..., decimals=3)`.  

I have tried to do that, but can't figuire yet where all changes have to be made in core/coords...  @fernandezc : your help would be more than welcome on this !

**Comment by fernandezc** [comment]
Yes this seems reasonable. 
May be the best would be that you make a pull request with the solution you tried, and I can have a look to the code!

FYI, I am currently rewriting a part of the core/dataset classes with, I hope, a  >90% coverage (I already find a lot of mistake!) and including the change related to the new `quantity` attribute (cf. our last discussions).  It is a lot of rewriting, but I think necessary to make the code more maintainable.
So It will be time to include this request as this rewriting will include LinearCoord.

**Comment by fernandezc** [comment]
Just to understand better what you need, let's  take an example (that I use in tests of linearcoord):
```python
    coord1 = LinearCoord.linspace(0.0, 60.0, 100, labels=None, units="minutes", title="time-on-stream" )
    coord2 = LinearCoord.linspace(4000.0, 1000.0, 100, labels=None, units="cm^-1", title="wavenumber")
```
The increments are respectively: 0.6060606060606061, and -30.303030303030305.
Do you mean that in the linespace calculation we should, change the limits (start and stop) so that the increment is rounded to e.g 3 decimals?
In the case of coord1 and coord2, stop would be: 59.99 and 1000.003 (If I count correctly :-)). 

One way if this is what is expected, it to use `arange` instead of `linspace`:
```python
coord1b = LinearCoord.arange( 0.0, 60.0, 0.606, labels=None, units="minutes", title="time-on-stream") 
```
The other situation is when we try to linearise existing data array. I will look at this.

**Reply by fernandezc** [reply]
The last PR should fix this problem. Tell me if it is OK for you (see test_coord_linearization for instance)

**Reply by fernandezc** [reply]
The only issue I imagine is that this change the stop of coord data series. May be we should share the difference with the actual data, by modifying equally start and stop. I am not sure to be clear enough.

**Reply by fernandezc** [reply]
start being the offset parameter.

**Comment by atravert** [comment]
Ok. Many thanks. I will try it on monday or tuesday the latest and i will tell you then !

Le 26 février 2022 15:16:28 GMT+01:00, Christian Fernandez ***@***.***> a écrit :
>The only issue I imagine is that this change the stop of coord data
>series. May be we should share the difference with the actual data, by
>modifying equally start and stop. I am not sure to be clear enough.
>
>—
>Reply to this email directly, view it on
>GitHub<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fspectrochempy%2Fspectrochempy%2Fdiscussions%2F398%23discussioncomment-2256583&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd7c84817cc494a72827908d9f9329556%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814817985439008%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=cpDamMHHxxcgOwkQiGOAnaZWStcWwHHRLwsaMnhd9Vk%3D&reserved=0>,
>or
>unsubscribe<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FACNBNCRFVE376YJQVQMU35DU5DODZANCNFSM5PKGV33Q&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd7c84817cc494a72827908d9f9329556%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814817985439008%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=HjGZobm9OJXv5PkCEl1bPtZjBP36l8GlmoDn8yFiTzQ%3D&reserved=0>.
>Triage notifications on the go with GitHub Mobile for
>iOS<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd7c84817cc494a72827908d9f9329556%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814817985439008%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Yf8NXmtFTmo%2FlNxi9PKaxt7mhDVECYkboWF3iHiuVnQ%3D&reserved=0>
>or
>Android<https://eur02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Carnaud.travert%40ensicaen.fr%7Cd7c84817cc494a72827908d9f9329556%7C430c3e75ff03445fb3f4ae2809a0b10b%7C0%7C0%7C637814817985439008%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=ZJHng6bcCYYH5h9IVCNFX9Y8r7HZe4dAxhgw%2F8Bo5mY%3D&reserved=0>.
>You are receiving this because you authored the thread.Message ID:
***@***.***>

-- Envoyé depuis /e/ Mail.
